wasm/src/main/scala/ir2wasm/WasmBuilder.scala

```diff
@@ -422,6 +418,368 @@ class WasmBuilder {
+    val superArgsFun = {
+      implicit val fctx = WasmFunctionContext(
+        Some(clazz.className),
+        WasmFunctionName.superArgs(clazz.className),
+        Some(jsClassCaptures),
+        Some(preSuperDecls),
+        hasNewTarget = true,
+        receiverTyp = None,
+        ctor.args,
+        List(WasmAnyRef) // a js.Array
+      )
+
+      WasmExpressionBuilder.generateIRBody(
+        IRTrees.JSArrayConstr(ctorBody.superCall.args),
+        IRTypes.AnyType
+      )
+
+      fctx.buildAndAddToContext()
+    }
+
+    // Build the `postSuperStats` function
+    val postSuperStatsFun = {
+      implicit val fctx = WasmFunctionContext(
+        Some(clazz.className),
+        WasmFunctionName.postSuperStats(clazz.className),
+        Some(jsClassCaptures),
+        Some(preSuperDecls),
+        hasNewTarget = true,
+        receiverTyp = Some(WasmAnyRef),
+        ctor.args,
+        List(WasmAnyRef)
+      )
+
+      import fctx.instrs
+
+      // Create fields
+      for (fieldDef <- clazz.fields if !fieldDef.flags.namespace.isStatic) {
+        // Load instance
+        instrs += LOCAL_GET(fctx.receiverStorage.idx)
+
+        // Load name
+        fieldDef match {
+          case IRTrees.FieldDef(_, name, _, _) =>
+            instrs += GLOBAL_GET(
+              GlobalIdx(WasmGlobalName.WasmGlobalJSPrivateFieldName(name.name))
+            )
+          case IRTrees.JSFieldDef(_, nameTree, _) =>
+            WasmExpressionBuilder.generateIRBody(nameTree, IRTypes.AnyType)
+        }
+
+        // Generate boxed representation of the zero of the field
+        WasmExpressionBuilder.generateIRBody(IRTypes.zeroOf(fieldDef.ftpe), IRTypes.AnyType)
+
+        instrs += CALL(FuncIdx(WasmFunctionName.installJSField))
+      }
+
+      WasmExpressionBuilder.generateIRBody(
+        IRTrees.Block(ctorBody.afterSuper),
+        IRTypes.AnyType
+      )
+
+      fctx.buildAndAddToContext()
+    }
+
+    // Build the actual `createJSClass` function
+    val createJSClassFun = {
+      implicit val fctx = WasmFunctionContext(
+        Some(clazz.className),
+        WasmFunctionName.createJSClassOf(clazz.className),
+        None,
+        None,
+        jsClassCaptures,
+        List(WasmRefType.any)
+      )
+
+      import fctx.instrs
+
+      // Bundle class captures in a capture data struct -- leave it on the stack for createJSClass
+      val dataStructType = ctx.getClosureDataStructType(jsClassCaptures.map(_.ptpe))
+      val dataStructLocal = fctx.addLocal(
+        "__classCaptures",
+        WasmRefType(WasmHeapType.Type(dataStructType.name))
+      )
+      for (cc <- jsClassCaptures)
+        instrs += LOCAL_GET(fctx.lookupLocalAssertLocalStorage(cc.name.name))
+      instrs += STRUCT_NEW(dataStructType.name)
+      instrs += LOCAL_TEE(dataStructLocal)
+
+      // Load super constructor
+      val jsSuperClassTree = clazz.jsSuperClass.getOrElse {
+        IRTrees.LoadJSConstructor(clazz.superClass.get.name)
```

Comment:
Because that is the spec 🤷‍♂️ 
https://lampwww.epfl.ch/~doeraene/sjsir-semantics/#sec-sjsir-classdef-runtime-semantics-evaluation

I added a comment with a link to the spec and a summary of what it says.

Reply:
Thanks!